### DISCUSSION TITLE
Is it possible to use up-mode="swap" and retain history

### DISCUSSION THREAD

**Discussion by jogaman** [unpoly/unpoly#395] [General]
Hi there,
I'm new to unpoly and loving it.

If i open a modal, and through that modal I go further bu using mode="swap", inside the same modal, then the history is lost. Is there a way I can keep the history?

1. / root
2. /modal1 opens as modal
3. inside modal, go to /modal2

if i click back i go directly to /, but I would like to go to /modal1 (ideally inside the modal). Is it possible?

**Comment by adam12** [comment]
I'm presuming you're referring to `[up-layer="swap"]` and not `[up-mode]`?

How are you visiting `/modal2` in the modal? And can you enable the console log and share the relevant entries? `up.log.enable()`

**Comment by jogaman** [comment]
Yes! All links have

`up-layer="swap" up-mode="modal" up-size="full"`

also between /modal1 and /modal2

full log / -> /modal1 -> /modal2 and then press back (I end up at / but is hoping to go to /modal1)

```
up.hello() Compiling fragment  
<body up-source="/">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<body up-source="/">
unpoly.min.js:1:116604
Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content. main.js:5:16
up:link:follow Following link unpoly.min.js:1:116604
up:request:load Loading GET /projects/p1/ unpoly.min.js:1:116604
up.request() Change with { solo } option will abort other requests unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET /projects/p1/ (14114 characters) unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET /projects/p1/ unpoly.min.js:1:116604
up.render() Could not match primary target ([up-main~=modal]). Updating a fallback target (main). unpoly.min.js:1:116604
up.render() Opening element "main" in new overlay unpoly.min.js:1:116604
up:layer:open Opening new overlay unpoly.min.js:1:116604
up:location:changed Advanced to location /projects/p1/ unpoly.min.js:1:116604
up.hello() Compiling fragment  
<up-modal size="full" nesting="0">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<up-modal size="full" nesting="0">
unpoly.min.js:1:116604
up:layer:opened Opened new modal overlay unpoly.min.js:1:116604
up:link:follow Following link unpoly.min.js:1:116604
up:request:load Loading GET /about/something/ unpoly.min.js:1:116604
up.request() Change with { solo } option will abort other requests unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET /about/something/ (70193 characters) unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET /about/something/ unpoly.min.js:1:116604
up.render() Could not match primary target ([up-main~=modal]). Updating a fallback target (main). unpoly.min.js:1:116604
up.render() Opening element "main" in new overlay unpoly.min.js:1:116604
up:layer:open Opening new overlay unpoly.min.js:1:116604
up:layer:dismiss Will dismiss modal overlay with value  :peel unpoly.min.js:1:116604
up:location:changed Advanced to location / unpoly.min.js:1:116604
up:layer:dismissed Dismissed modal overlay with value  :peel unpoly.min.js:1:116604
up:location:changed Advanced to location /about/something/ unpoly.min.js:1:116604
up.hello() Compiling fragment  
<up-modal size="full" nesting="0">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<up-modal size="full" nesting="0">
unpoly.min.js:1:116604
up:layer:opened Opened new modal overlay unpoly.min.js:1:116604
up:request:load Loading GET / unpoly.min.js:1:116604
up:request:loaded Server responded HTTP 200 to GET / (132747 characters) unpoly.min.js:1:116604
up:fragment:loaded Loaded fragment from successful response to GET / unpoly.min.js:1:116604
up.render() Updating "body" in root layer unpoly.min.js:1:116604
up:layer:dismiss Will dismiss modal overlay with value  :peel unpoly.min.js:1:116604
up:layer:dismissed Dismissed modal overlay with value  :peel unpoly.min.js:1:116604
up.hello() Compiling fragment  
<body up-source="/">
unpoly.min.js:1:116604
up:fragment:inserted Inserted fragment  
<body up-source="/">
unpoly.min.js:1:116604
up.viewport.restoreScroll() Restoring scroll positions for URL / to  
Object {  }
unpoly.min.js:1:116604
up:location:changed Restored location / unpoly.min.js:1:116604

​


```

**Reply by adam12** [reply]
The history is kind of hard to follow. Can you confirm? I based this on `up:location:changed` events in your log.

```
/
/projects/one
/
/about/something 
/
```

**Reply by jogaman** [reply]
Hi yes, i tried to redact out the unique urls above, but yes. Sorry for the extra work, would you mind changing the urls above as well?

Should be like

```
1. /
2. /project/p1
3. /about/something/
4. /
```
There shouldn't be a / in between 2 and 3. At least not intentionally, or anything i can perceive.

**Comment by jogaman** [comment]
Ahaaa! Indeed there is an extra / in between the pages! (between 2 and 3)
Not sure what it means, but there is a quick redirect to / in between the two modals!

The history is there, there is just an extra / in the history

**Reply by adam12** [reply]
Could you take the [Unpoly glitch template](https://glitch.com/edit/#!/empty-unpoly-project?path=index.html%3A1%3A0) and see if you can create a simple reproduction? It might help us look a bit closer.

**Reply by jogaman** [reply]
For sure, here it is https://dolomite-scarlet-cornucopia.glitch.me/

**Reply by adam12** [reply]
That helped. It kind of looks like a bug, since it actually doesn't make the request to `/`. Maybe Henning knows.

**Comment by jogaman** [comment]
Should i make an issue?

And, would you mind redacting the urls above? (I was too fast pasting the original logs, sorry)

**Reply by adam12** [reply]
Yes, I'd suggest making an issue.

I've redacted the URLs.

**Reply by jogaman** [reply]
Thank you!
